akka-stream/src/main/scala/akka/stream/impl/AcknowledgeSubscriber.scala

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.actor.{ ActorLogging, ActorRef, Props, Status }
+import akka.stream.actor.ActorPublisherMessage.Request
+import akka.stream.actor.{ ActorSubscriber, ActorSubscriberMessage, RequestStrategy }
+
+import scala.util.{ Try, Failure, Success }
+
+private[akka] object AcknowledgeSubscriber {
+  def props(highWatermark: Int) =
+    Props(new AcknowledgeSubscriber(highWatermark))
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class AcknowledgeSubscriber(maxBuffer: Int) extends ActorSubscriber with ActorLogging {
+  import ActorSubscriberMessage._
+
+  var buffer: Vector[Any] = Vector.empty
+
+  override val requestStrategy = new RequestStrategy {
+    def requestDemand(remainingRequested: Int): Int = {
+      maxBuffer - buffer.size - remainingRequested
+    }
+  }
+
+  var requester: Option[ActorRef] = None
+
+  def receive = {
+    case Request(_) ⇒
+      if (requester.isEmpty) {
+        requester = Some(sender)
+        trySendElementDownstream()
+      } else
+        sender ! Status.Failure(
+          new IllegalStateException("You have to wait for first future to be resolved to send another request"))
+
+    case OnNext(elem) ⇒
+      if (maxBuffer != 0) {
+        buffer :+= elem
+        trySendElementDownstream()
+      } else requester match {
+        case Some(ref) ⇒
+          requester = None
+          ref ! Some(elem)
+        case None ⇒ log.debug("Dropping element because there is no downstream demand: [{}]", elem)
```

Comment:
this line works when buffer size is 0 so we drop everything if there is no request waiting for element